File: implementations/typescript/packages/mcp-server/src/stdio.ts
// SPDX-License-Identifier: Apache-2.0
import type { Readable, Writable } from "node:stream";
import { once } from "node:events";
import { canonicalJson } from "@psp-cdl/core";
import { MAX_REQUEST_BYTES } from "@psp-cdl/api-server";
import { McpServer } from "./server.js";
/** Bounded newline framing; stdout carries protocol messages only. Oversized/truncated frames close the connection. */
export async function serveStdio(server:McpServer,input:Readable=process.stdin,output:Writable=process.stdout):Promise<void> {
  let pending=Buffer.alloc(0);
  for await(const chunk of input) {
    const bytes=Buffer.from(chunk);
    let start=0;
    while(start<bytes.length) {
      const newline=bytes.indexOf(10,start), end=newline<0?bytes.length:newline;
      if(pending.length+end-start>MAX_REQUEST_BYTES) throw new Error("FRAME_TOO_LARGE");
      pending=Buffer.concat([pending,bytes.subarray(start,end)]);
      if(newline<0) break;
      const source=new TextDecoder("utf-8",{fatal:true}).decode(pending);
      pending=Buffer.alloc(0);start=newline+1;
      const reply=await server.handle(source);
      if(reply!==null&&!output.write(canonicalJson(reply)+"\n")) await once(output,"drain");
    }
  }
  if(pending.length) throw new Error("TRUNCATED_FRAME");
}

File: implementations/python/packages/mcp-server/src/psp_cdl_mcp_server/stdio.py
# SPDX-License-Identifier: Apache-2.0
import sys
from psp_cdl_core import canonical_json
from psp_cdl_api_server import MAX_REQUEST_BYTES

def serve_stdio(server,input_stream=None,output_stream=None):
    """Bounded binary newline framing; no protocol output for notifications."""
    source=input_stream if input_stream is not None else sys.stdin.buffer
    target=output_stream if output_stream is not None else sys.stdout.buffer
    while True:
        frame=source.readline(MAX_REQUEST_BYTES+2)
        if not frame:
            return
        if len(frame.rstrip(b"\n"))>MAX_REQUEST_BYTES:
            raise ValueError("FRAME_TOO_LARGE")
        if not frame.endswith(b"\n"):
            raise ValueError("TRUNCATED_FRAME")
        reply=server.handle(frame[:-1].decode("utf-8",errors="strict"))
        if reply is not None:
            wire=canonical_json(reply).encode("utf-8")
            if len(wire)>MAX_REQUEST_BYTES:
                raise ValueError("FRAME_TOO_LARGE")
            target.write(wire+b"\n")
            target.flush()
